File: .a-new-structure/b-word/L0-universal/hybrid/foundation/types/workflow.go
//omni:code --go -library
//omni:key B-L0-foundation-types-workflow
//omni:version a-01.00

// ============================================================================
// METADATA
// ============================================================================
//
// Key: B-L0-foundation-types-workflow
// Purpose: Universal workflow operation types for step-based execution
// Biblical: Proverbs 16:9 - "A man's heart deviseth his way"
// Authors: Nova Dawn
// Version: 1.0.0
// Created: 2026-02-16
//
// Extracted from L3-cpisi types.go — these are generic workflow tracking
// types usable by any layer. CPI-SI specific types (Choice, Halt, Tendency,
// Hebrew states, K:ALIGN) remain in L3.
//
// ============================================================================

package types

// ============================================================================
// SETUP
// ============================================================================

// WorkflowOperation represents the type of operation being performed.
// Maps to workflow/*.jsonc files.
type WorkflowOperation string

const (
	WorkflowNone         WorkflowOperation = ""              // No active workflow
	WorkflowFileCreation WorkflowOperation = "file-creation" // Creating new files from templates
	WorkflowAlignment    WorkflowOperation = "alignment"     // Aligning documentation with reality
	WorkflowBlockUpdate  WorkflowOperation = "block-update"  // Updating existing files block by block
	WorkflowCommit       WorkflowOperation = "commit"        // Git commit workflow
)

// WorkflowStep represents a step within a workflow
type WorkflowStep struct {
	Index       int    `json:"index"`       // Step number (1-based)
	Action      string `json:"action"`      // What to do
	Description string `json:"description"` // Human-readable description
	Completed   bool   `json:"completed"`   // Whether step is done
}

// RuntimeWorkflow tracks current workflow state
type RuntimeWorkflow struct {
	Operation   WorkflowOperation `json:"operation"`    // Current operation type
	CurrentStep int               `json:"current_step"` // Current step index (1-based)
	TotalSteps  int               `json:"total_steps"`  // Total steps in workflow
	StartedAt   string            `json:"started_at"`   // When workflow started
	Steps       []WorkflowStep    `json:"steps"`        // All steps with completion status
}

// ============================================================================
// BODY
// ============================================================================

// IsActive returns whether a workflow is currently in progress
func (w *RuntimeWorkflow) IsActive() bool {
	return w.Operation != WorkflowNone
}

// IsComplete returns whether all steps are done
func (w *RuntimeWorkflow) IsComplete() bool {
	return w.CurrentStep > w.TotalSteps && w.TotalSteps > 0
}

// Progress returns completion percentage (0-100)
func (w *RuntimeWorkflow) Progress() int {
	if w.TotalSteps == 0 {
		return 0
	}
	completed := 0
	for _, s := range w.Steps {
		if s.Completed {
			completed++
		}
	}
	return completed * 100 / w.TotalSteps
}

// ============================================================================
// CLOSING
// ============================================================================
// Generic workflow types — extracted from L3, usable by all layers.

File: .a-new-structure/b-word/L0-universal/c-hybrid/foundation/src/workflow.rs
//omni:code --rust -library
//omni:key B-L0-foundation-workflow-rs
//omni:version a-01.00

// ============================================================================
// METADATA
// ============================================================================

//! # Workflow Types (Rust)
//!
//! Universal workflow operation types for step-based execution.
//!
//! Extracted from L3-cpisi — these are generic workflow tracking types
//! usable by any layer. CPI-SI specific types remain in L3.

// ============================================================================
// SETUP
// ============================================================================

use serde::{Deserialize, Serialize};
use std::fmt;

// ────────────────────────────────────────────────────────────────
// S.1 Types
// ────────────────────────────────────────────────────────────────

/// Type of workflow operation being performed.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub enum WorkflowOperation {
    /// No active workflow.
    #[serde(rename = "")]
    #[default]
    None,
    /// Creating new files from templates.
    #[serde(rename = "file-creation")]
    FileCreation,
    /// Aligning documentation with reality.
    #[serde(rename = "alignment")]
    Alignment,
    /// Updating existing files block by block.
    #[serde(rename = "block-update")]
    BlockUpdate,
    /// Git commit workflow.
    #[serde(rename = "commit")]
    Commit,
}

/// A step within a workflow.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowStep {
    /// Step number (1-based).
    pub index: usize,
    /// What to do.
    pub action: String,
    /// Human-readable description.
    pub description: String,
    /// Whether step is done.
    pub completed: bool,
}

/// Tracks current workflow state.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuntimeWorkflow {
    /// Current operation type.
    pub operation: WorkflowOperation,
    /// Current step index (1-based).
    pub current_step: usize,
    /// Total steps in workflow.
    pub total_steps: usize,
    /// When workflow started.
    #[serde(default)]
    pub started_at: String,
    /// All steps with completion status.
    #[serde(default)]
    pub steps: Vec<WorkflowStep>,
}

// ============================================================================
// BODY
// ============================================================================

// ────────────────────────────────────────────────────────────────
// B.1 WorkflowOperation
// ────────────────────────────────────────────────────────────────

// Default derived on enum declaration above (None variant)

impl fmt::Display for WorkflowOperation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::None => write!(f, "none"),
            Self::FileCreation => write!(f, "file-creation"),
            Self::Alignment => write!(f, "alignment"),
            Self::BlockUpdate => write!(f, "block-update"),
            Self::Commit => write!(f, "commit"),
        }
    }
}

// ────────────────────────────────────────────────────────────────
// B.2 RuntimeWorkflow
// ────────────────────────────────────────────────────────────────

impl RuntimeWorkflow {
    /// Whether a workflow is currently in progress.
    pub fn is_active(&self) -> bool {
        self.operation != WorkflowOperation::None
    }

    /// Whether all steps are done.
    pub fn is_complete(&self) -> bool {
        self.total_steps > 0 && self.current_step > self.total_steps
    }

    /// Completion percentage (0-100).
    pub fn progress(&self) -> usize {
        if self.total_steps == 0 {
            return 0;
        }
        let completed = self.steps.iter().filter(|s| s.completed).count();
        completed * 100 / self.total_steps
    }
}

impl Default for RuntimeWorkflow {
    fn default() -> Self {
        Self {
            operation: WorkflowOperation::None,
            current_step: 0,
            total_steps: 0,
            started_at: String::new(),
            steps: Vec::new(),
        }
    }
}

// ============================================================================
// CLOSING
// ============================================================================
//
// Generic workflow types — extracted from L3, usable by all layers.
// "A man's heart deviseth his way" — Proverbs 16:9

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_inactive() {
        let w = RuntimeWorkflow::default();
        assert!(!w.is_active());
        assert!(!w.is_complete());
        assert_eq!(w.progress(), 0);
    }

    #[test]
    fn progress_calculation() {
        let w = RuntimeWorkflow {
            operation: WorkflowOperation::FileCreation,
            current_step: 2,
            total_steps: 4,
            started_at: String::new(),
            steps: vec![
                WorkflowStep { index: 1, action: "a".into(), description: "".into(), completed: true },
                WorkflowStep { index: 2, action: "b".into(), description: "".into(), completed: true },
                WorkflowStep { index: 3, action: "c".into(), description: "".into(), completed: false },
                WorkflowStep { index: 4, action: "d".into(), description: "".into(), completed: false },
            ],
        };
        assert!(w.is_active());
        assert!(!w.is_complete());
        assert_eq!(w.progress(), 50);
    }

    #[test]
    fn operation_display() {
        assert_eq!(WorkflowOperation::Commit.to_string(), "commit");
        assert_eq!(WorkflowOperation::None.to_string(), "none");
    }
}
